Improve and fix port hotzones of GraphNode

Geometror opened the pull request:
Split from #61414 with additional fixes. Based on the approach in #49722.
Superceeds #49722.

![port_hotzones_improvements](https://user-images.githubusercontent.com/50084500/170847719-f6c3fd04-ad89-4910-bc55-f73b4a5ef44e.gif) ![port_hotzones_improvements2](https://user-images.githubusercontent.com/50084500/170847721-95e0c2ca-2763-4c02-8e98-5a92520fbbce.gif)


Detailed changes:
- The hotzone of each port uses the actual height of the slot instead of a constant distance
- Fix sub-child control filtering (seems like that code has been broken since the introduction of zooming in GraphEdit), which removes the problems mentioned in #49722 without introducing two hotzone "modes"
- Small code style/structural improvements

fire reviewed (approved):
Looks good.

I downloaded the editor, tested the hotspot and it seemed as expected.

Didn't see any obvious coding errors.

kleonc commented:
I see in #49722 that hotzones expand both ways horizontally, the same in this PR. But they were made to expand outwards only in #51517. Seems like regression to me. It was introduced in #52015 (`GraphEdit::is_in_port_hotzone` was made to ignore its `p_left` parameter). Could be probably fixed in here?

fire commented:
I think this implementation feels better. Are you saying we should change it to be https://github.com/godotengine/godot/pull/52015?

kleonc commented:
> Are you saying we should change it to be #52015?

No, I'm saying improvement made in #51517 was removed by #52015. So it could potentially be reintroduced here.

![firefox_BunlrYPUcu](https://user-images.githubusercontent.com/9283098/170873415-cce0bd6f-6764-44c7-af01-6c3dbac1fa54.png)
(overlay drawn manually, a little exaggerated sizing)

Geometror commented:
I think it should be fine like that, since the sub-control filtering code works again (which eliminates the problems that were the motivation for the changes in https://github.com/godotengine/godot/pull/51517, the hotzones overlapped and blocked some clickable controls in GraphNodes). However, I'm not against exposing two theme constants (for inner and outer port grab distance) to have more control when tweaking the hotzones.

kleonc commented:
@Geometror If it works correctly and issues like #47288 will not arise again then I guess it's fine. I just got surprised looking at the changes here as I recalled the change from #51517.

So if the hotspots are meant to extend horizontally in both ways then I guess `bool p_left` parameter of `GraphEdit::is_in_port_hotzone` is no longer needed and could be removed (as already mentioned it's unused since #52015).

Also cc @Chaosus to take a look.

Also to point it out: in `3.x` it still works like in #51517 as #52015 wasn't backported. So there's no regression in there.

Chaosus commented:
Looks good for me (a bit tested it locally). I would add some port highlighting to tell the user where he/she may connect it while dragging.

Geometror commented:
Ok, I removed `"port_grab_distance_horizontal"` and added `"port_hotzone_inner_extent"` and `"port_hotzone_outer_extent"`, to allow for more granular tweaking. (uses the p_left parameter again)
@Chaosus I think port highlighting should go in another PR to keep this straight-forward for reviewing.

Geometror committed:
Improve and fix the GraphNode port hotzones

Co-authored-by: Ansraer <jacky2611@gmail.com>

Chaosus commented:
@Geometror doc need to be updated

Geometror commented:
Updated the docs, should be ready to merge :)

Chaosus commented:
Thanks!
